### ISSUE TITLE
Use SSE from THVector

### ISSUE THREAD

**Issue by fmassa** [torch/torch7#323]
I've noticed that we have SSE instructions for some basic operations in https://github.com/torch/torch7/blob/master/lib/TH/THVector.h .

Is there a reason why `THVector` is not used in `TH` math operations (like `add`, `mul`, etc), at least for the contiguous case, in addition to the `omp` directives ?

**Comment by soumith** [comment]
I think it is mostly because no one spent time doing it. THVector stuff only applies to 1D (contiguous) case usually (you can extend to 2D if it is contiguous for example)

**Comment by fmassa** [comment]
I don't know if it would bring any improvements actually.

I played a bit with it in https://github.com/fmassa/torch-nn/tree/master/sse_optim for the `torch.mul` function using `THDoubleVector_scale` function and the micro-benchmarks I did didn't show any significant improvements over `OMP` (but maybe I did something wrong ?).
Also, the `-O3` flag seems to remove all the benefits of using `THVector`.

But then, those micro-bechmarks are almost always flawed :)

**Comment by soumith** [comment]
sse and openmp are independent optimizations. one can do OpenMP + SSE for example.

**Comment by fmassa** [comment]
@soumith I did exactly that in my previous link, maybe I should have chosen better names.
In my small tests, `parsseMul` stands for OpenMP + SSE https://github.com/fmassa/torch-nn/blob/master/sse_optim/sse.c#L38 , and there was not much gain in doing both compared to doing only OpenMP. But maybe I did something wrong ?

**Comment by andreaskoepf** [comment]
General questing regarding use of OpenMP: Lots of code uses the [TH_TENSOR_APPLY](https://github.com/torch/torch7/blob/e271689b897780b4e4042641a20f4885931fa185/lib/TH/THTensorApply.h) methods but they seem not to have any 'open-mp over threshold path' as for example [THTensorMath.c](https://github.com/torch/torch7/blob/086ec2a760a15a5165b488e25899f3ee70e2ec6e/lib/TH/generic/THTensorMath.c) has. Most of the TH_TENSOR_APPLY using code I saw so far was effectively side-effect free and could immediately use slightly modified 'parallel for' TH_TENSOR_APPLY methods.

**Comment by dominikgrewe** [comment]
I guess there are at least to considerations for using OpenMP:

1) The output tensors must be "non-overlapping", so that no element is written twice (though you could argue that this case is somewhat undefined anyway). Currently the OpenMP versions simply approximate that with checking that tensors are contiguous, which also simplifies index computations.

2) You must not break out of the loop. Some uses of TH_TENSOR_APPLY error out in some conditions which isn't allowed in OpenMP loops I believe.

I think a "TH_TENSOR_APPLY_PARALLEL" could be useful and increase the use of OpenMP within Torch. The current implementation of TH_TENSOR_APPLY doesn't lend itself to parallelisation very easily I think, because it's main loop is a while loop. But I suppose this can be changed.

**Comment by fmassa** [comment]
@andreaskoepf I agree with you. But reading the history of `TH` commits, it seems that @clementfarabet tried that in the past but that wasn't trivial, see the commit message from https://github.com/torch/torch7/commit/1864e3933be1f4010b262ab60a2c3f221b1801c9
The reason why it's not trivial is maybe but to the fact that it's not possible to use preprocessor directives inside a macro, and `TH_TENSOR_APPLY` is indeed a macro.
